src/hpx_init.cpp

```diff
@@ -340,6 +340,53 @@ namespace hpx
             hpx::util::may_attach_debugger("test-failure");
         }
 
+#if defined(HPX_HAVE_VERIFY_LOCKS)
+        void registered_locks_error_handler()
+        {
+            std::string back_trace = hpx::util::trace(std::size_t(128));
+
+            // throw or log, depending on config options
+            if (get_config_entry("hpx.throw_on_held_lock", "1") == "0")
+            {
+                if (back_trace.empty())
+                {
+                    LERR_(debug)
+                        << "suspending thread while at least one lock is "
+                           "being held (stack backtrace was disabled at "
+                           "compile time)";
+                }
+                else
+                {
+                    LERR_(debug)
+                        << "suspending thread while at least one lock is "
+                        << "being held, stack backtrace: " << back_trace;
+                }
+            }
+            else
+            {
+                if (back_trace.empty())
+                {
+                    HPX_THROW_EXCEPTION(invalid_status, "verify_no_locks",
+                        "suspending thread while at least one lock is "
+                        "being held (stack backtrace was disabled at "
+                        "compile time)");
+                }
+                else
+                {
+                    HPX_THROW_EXCEPTION(invalid_status, "verify_no_locks",
+                        "suspending thread while at least one lock is "
+                        "being held, stack backtrace: " +
+                            back_trace);
+                }
+            }
+        }
+
+        bool register_locks_predicate()
+        {
+            return threads::get_self_ptr() != nullptr;
+        }
+#endif
```

Comment:
Does this code needs to live in hpx_init.cpp (i.e. does it need to be linked to the executable)? Wouldn't it be ok to have it exposed from the main core library?

Reply:
No, and yes, I think. All of the handlers above could go in a separate file (just like `termination_handler` in fact).

Reply:
I agree. This file should be part of HPX core, however, not HPX init.

Reply:
Yep, that's what I understood from your first comment, although my reply might have been a bit unclear.

Reply:
I'll do it in a separate PR :)